File: Packages/OverlayKit/Tests/OverlayKitTests/PinGeometryTests.swift
import TestKit
import CoreGraphics
@testable import OverlayKit

let pinGeometryTests: [TestCase] = [
    TestCase("retinaImageGetsPointSize") { t in
        let f = PinGeometry.initialFrame(
            imagePixelSize: CGSize(width: 400, height: 200), backingScale: 2,
            visibleFrame: CGRect(x: 0, y: 0, width: 1440, height: 875), sourceRect: nil)
        t.equal(f.size, CGSize(width: 200, height: 100))
    },
    TestCase("centersOnVisibleFrameWithoutSource") { t in
        let vf = CGRect(x: 0, y: 0, width: 1440, height: 875)
        let f = PinGeometry.initialFrame(imagePixelSize: CGSize(width: 400, height: 200),
                                         backingScale: 2, visibleFrame: vf, sourceRect: nil)
        t.approxEqual(f.midX, vf.midX)
        t.approxEqual(f.midY, vf.midY)
    },
    TestCase("centersOnSourceRect") { t in
        let f = PinGeometry.initialFrame(
            imagePixelSize: CGSize(width: 200, height: 100), backingScale: 2,
            visibleFrame: CGRect(x: 0, y: 0, width: 1440, height: 875),
            sourceRect: CGRect(x: 300, y: 300, width: 100, height: 50))
        t.approxEqual(f.midX, 350)
        t.approxEqual(f.midY, 325)
    },
    TestCase("clampsTo80PercentOfScreen") { t in
        let vf = CGRect(x: 0, y: 0, width: 1000, height: 800)
        let f = PinGeometry.initialFrame(imagePixelSize: CGSize(width: 4000, height: 2000),
                                         backingScale: 1, visibleFrame: vf, sourceRect: nil)
        t.approxEqual(f.width, 800)    // 80% of 1000 wide
        t.approxEqual(f.height, 400)   // aspect preserved
    },
    TestCase("staysInsideVisibleFrame") { t in
        let vf = CGRect(x: 0, y: 0, width: 1440, height: 875)
        let f = PinGeometry.initialFrame(
            imagePixelSize: CGSize(width: 400, height: 200), backingScale: 2,
            visibleFrame: vf,
            sourceRect: CGRect(x: 1400, y: 850, width: 100, height: 50))
        t.isTrue(vf.contains(f), "frame \(f) escapes \(vf)")
    },
    TestCase("zoomScalesAroundCenter") { t in
        let current = CGRect(x: 100, y: 100, width: 200, height: 100)
        let f = PinGeometry.zoomedFrame(current: current,
                                        naturalSize: CGSize(width: 200, height: 100),
                                        factor: 2)
        t.equal(f.size, CGSize(width: 400, height: 200))
        t.approxEqual(f.midX, current.midX)
        t.approxEqual(f.midY, current.midY)
    },
    TestCase("zoomClampsToMinAndMax") { t in
        let natural = CGSize(width: 200, height: 100)
        let current = CGRect(x: 0, y: 0, width: 200, height: 100)
        let big = PinGeometry.zoomedFrame(current: current, naturalSize: natural, factor: 100)
        t.equal(big.size, CGSize(width: 600, height: 300))      // 3× cap
        let small = PinGeometry.zoomedFrame(current: current, naturalSize: natural, factor: 0.001)
        t.equal(small.size, CGSize(width: 50, height: 25))      // 0.25× floor
    },
]

File: windows/tests/BetterScreenshot.Tests/PinGeometryTests.cs
using BetterScreenshot.Capture;
using BetterScreenshot.Core;
using Xunit;

namespace BetterScreenshot.Tests;

public class PinGeometryTests
{
    [Fact]
    public void RetinaImageGetsPointSize()
    {
        var f = PinGeometry.InitialFrame(new PxSize(400, 200), dpiScale: 2, new PxRect(0, 0, 2000, 2000));
        Assert.Equal(200, f.Width, 6);
        Assert.Equal(100, f.Height, 6);
    }

    [Fact]
    public void CentersOnVisibleFrameWithoutSource()
    {
        var vf = new PxRect(0, 0, 1000, 800);
        var f = PinGeometry.InitialFrame(new PxSize(200, 100), 1, vf);
        Assert.Equal(vf.Center.X, f.Center.X, 6);
        Assert.Equal(vf.Center.Y, f.Center.Y, 6);
    }

    [Fact]
    public void CentersOnSourceRect()
    {
        var vf = new PxRect(0, 0, 1000, 800);
        var f = PinGeometry.InitialFrame(new PxSize(100, 100), 1, vf, new PxRect(400, 300, 50, 50));
        Assert.Equal(425, f.Center.X, 6);
        Assert.Equal(325, f.Center.Y, 6);
    }

    [Fact]
    public void ClampsTo80PercentOfScreen()
    {
        var f = PinGeometry.InitialFrame(new PxSize(4000, 2000), 1, new PxRect(0, 0, 1000, 800));
        Assert.Equal(800, f.Width, 6);
        Assert.Equal(400, f.Height, 6);
    }

    [Fact]
    public void ClampsHeightLimitedImages()
    {
        var f = PinGeometry.InitialFrame(new PxSize(1000, 2000), 1, new PxRect(0, 0, 1000, 800));
        Assert.Equal(640, f.Height, 6); // 80% of 800
        Assert.Equal(320, f.Width, 6);
    }

    [Fact]
    public void StaysInsideVisibleFrame()
    {
        var vf = new PxRect(0, 0, 1000, 800);
        var f = PinGeometry.InitialFrame(new PxSize(400, 400), 1, vf, new PxRect(980, 780, 10, 10));
        Assert.True(f.X >= vf.X);
        Assert.True(f.Y >= vf.Y);
        Assert.True(f.Right <= vf.Right + 1e-6);
        Assert.True(f.Bottom <= vf.Bottom + 1e-6);
    }

    [Fact]
    public void NonZeroOriginVisibleFrameStillContains()
    {
        var vf = new PxRect(1440, 0, 1920, 1080);
        var f = PinGeometry.InitialFrame(new PxSize(800, 600), 1, vf);
        Assert.True(f.X >= vf.X);
        Assert.True(f.Y >= vf.Y);
        Assert.True(f.Right <= vf.Right + 1e-6);
        Assert.True(f.Bottom <= vf.Bottom + 1e-6);
    }

    [Fact]
    public void ZoomScalesAroundCenter()
    {
        var cur = new PxRect(100, 100, 200, 100);
        var f = PinGeometry.ZoomedFrame(cur, new PxSize(200, 100), 2);
        Assert.Equal(400, f.Width, 6);
        Assert.Equal(200, f.Height, 6);
        Assert.Equal(cur.Center.X, f.Center.X, 6);
        Assert.Equal(cur.Center.Y, f.Center.Y, 6);
    }

    [Fact]
    public void ZoomClampsToMinAndMax()
    {
        var cur = new PxRect(0, 0, 200, 100); // scale 1.0
        Assert.Equal(600, PinGeometry.ZoomedFrame(cur, new PxSize(200, 100), 100).Width, 6); // clamp 3.0
        Assert.Equal(50, PinGeometry.ZoomedFrame(cur, new PxSize(200, 100), 0.001).Width, 6); // clamp 0.25
    }
}
